data_model/primitives/src/conststr.rs

```diff
@@ -0,0 +1,670 @@
+    fn from(value: T) -> Self {
+        match value.try_into() {
+            Ok(inlined) => Self { inlined },
+            Err(value) => Self {
+                boxed: ManuallyDrop::new(value.into()),
+            },
+        }
+    }
+}
+
+impl Clone for ConstStringData {
+    fn clone(&self) -> Self {
+        if self.is_inlined() {
+            Self {
+                inlined: *self.inlined(),
+            }
+        } else {
+            Self {
+                boxed: self.boxed().clone(),
+            }
+        }
+    }
+}
+
+impl Drop for ConstStringData {
+    #[allow(unsafe_code)]
+    fn drop(&mut self) {
+        if !self.is_inlined() {
+            // SAFETY: safe because`is_inlined` == `false`.
+            unsafe {
+                ManuallyDrop::drop(&mut self.boxed);
+            }
+        }
+    }
+}
+
+#[derive(DebugCustom)]
+#[debug(fmt = "{:?}", "&**self")]
+#[repr(C)]
+struct BoxedString {
+    #[cfg(target_endian = "little")]
+    ptr: *const u8,
+    len: usize,
+    #[cfg(target_endian = "big")]
+    ptr: *const u8,
+}
+
+impl BoxedString {
+    #[inline]
+    const fn len(&self) -> usize {
+        self.len
+    }
+
+    #[allow(unsafe_code)]
+    #[inline]
+    fn as_bytes(&self) -> &[u8] {
+        // SAFETY: valid allocation created from `Box` or cloned.
+        unsafe { from_raw_parts(self.ptr, self.len) }
+    }
+
+    #[inline]
+    fn layout(len: usize) -> Layout {
+        if let Ok(layout) = Layout::array::<u8>(len) {
+            layout
+        } else {
+            Self::overflow()
+        }
+    }
+
+    /// There is little that can be done if an allocation fails
+    /// due to incorrect layout or memory exhaustion.
+    #[allow(clippy::panic)]
+    #[inline]
+    const fn overflow() -> ! {
+        panic!("overflow");
+    }
+}
+
+impl AsRef<str> for BoxedString {
+    #[allow(unsafe_code)]
+    #[inline]
+    fn as_ref(&self) -> &str {
+        // SAFETY: created from valid utf-8
+        unsafe { from_utf8_unchecked(self.as_bytes()) }
+    }
+}
+
+impl Deref for BoxedString {
+    type Target = str;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        self.as_ref()
+    }
+}
+
+impl Clone for BoxedString {
+    /// Properly clone `BoxedString` into new allocation.
+    #[allow(unsafe_code)]
+    fn clone(&self) -> Self {
+        let len = self.len;
+        let layout = Self::layout(len);
+        // SAFETY: pointer is checked to be non-null.
+        let ptr = unsafe { alloc::alloc::alloc(layout) };
+        if ptr.is_null() {
+            Self::overflow();
+        }
+        // SAFETY: all conditions are met.
+        unsafe { ptr.copy_from_nonoverlapping(self.ptr, self.len) };
+        Self { ptr, len }
+    }
+}
+
+impl From<&str> for BoxedString {
+    #[allow(unsafe_code)]
+    #[inline]
+    fn from(value: &str) -> Self {
+        let len = value.len();
+        let boxed = value.as_bytes().to_vec().into_boxed_slice();
+        let ptr = Box::into_raw(boxed).cast::<u8>();
```

Comment:
the same, this cast is most likely a UB. I've checked the size of this pointer and it's a fat pointer you are casting to a raw pointer